Improve Logging Example and Documentation (#7424)
microsoft/FluidFramework: docs/content/docs/testing/telemetry.md

## Collect Fluid Framework logs with a custom `ITelemetryBaseLogger`

The `ITelemetryBaseLogger` is an interface within the `@fluidframework/common-definitions` package. This interface can
be implemented and passed into the client's constructor via the props parameter.

> All Fluid service clients (for example, `AzureClientProps` and `TinyliciousClientProps`) allow passing a `logger?: ITelemetryBaseLogger`

Reviewer: ```suggestion
All Fluid service clients (for example, `AzureClient` and `TinyliciousClient`) allow passing a `logger?: ITelemetryBaseLogger`
```

into the service client props. Both `createContainer()` and `getContainer()` methods will
create an instance of a `Loader` class object, where the logger defined in the service client props is passed in

> as an optional parameter, `ILoaderProps.logger`, in the `Loader` constructor argument.

Reviewer: ```suggestion
as an optional parameter, `ILoaderProps.logger`, to the `Loader` constructor.
```

> [TinyliciousClientProps](https://github.com/microsoft/FluidFramework/blob/main/packages/framework/tinylicious-client/src/interfaces.ts#L17)

Reviewer: linking to the code like this doesn't seem maintainable.  is this definition in the doc site api reference somewhere?  it would be better to link there instead if you can because it's more stable, for this and elsewhere

Author: I couldn't find anything on the official website [here](https://fluidframework.com/docs/testing/tinylicious/), I think it is yet to be updated or written. The closest thing I found that mentioned about the props is [here](https://github.com/microsoft/FluidFramework/tree/main/packages/framework/tinylicious-client#instantiating-tinyliciousclient)

Reviewer 2: We should link to the package docs, but they're not published right now. Once #7472 is decided, we can add appropriate links. I recommend removing the links for now and following up to add them later.

Author: Right, I will create a follow-up Github issue for this. 😊

Author: Github issue created #7480

> interface definition takes an optional parameter `logger`. (The definition is similar to `AzureClientProps` interface)

Reviewer 3: nit: include link to `AzureClientProps` here, similar to Tiny?

Author: good call. Added the link

[createContainer()](https://github.com/microsoft/FluidFramework/blob/main/packages/framework/tinylicious-client/src/TinyliciousClient.ts#L56)

> interface definition takes a `TinyliciousClientProps` type as its `serviceContainerConfig` argument. (The

Reviewer 3: This is no longer the case right? It just takes in the `containerSchema`, nothing else

Author: right, I have removed both `createContainer` and `getContainer` and only have the props mentioned

`AzureClient` will take `AzureContainerConfig` for its respective methods)

[getContainer()](https://github.com/microsoft/FluidFramework/blob/main/packages/framework/tinylicious-client/src/TinyliciousClient.ts#L72)

> interface definition takes a `TinyliciousContainerConfig` type as its `serviceContainerConfig` argument. (The

Reviewer 3: ```suggestion
interface definition takes the `id` of the container and its `containerSchema`. (The
```

Author: marking it as resolved since `getContainer` is no longer relevant and taken out

> `AzureClient` will take `AzureContainerConfig` for its respective methods)

Reviewer: this isn't true anymore.  can probably be removed since it's no longer relevant

## Collect Fluid Framework logs with a custom `ITelemetryBaseLogger`

The `ITelemetryBaseLogger` is an interface within the `@fluidframework/common-definitions` package. This interface can

> be implemented and passed into the client's constructor via the props parameter.

Reviewer 2: ```suggestion
be implemented and passed into the service client's constructor via the `props` parameter.
```

All Fluid service clients (for example, `AzureClient` and `TinyliciousClient`) allow passing a `logger?: ITelemetryBaseLogger`
into the service client props. Both `createContainer()` and `getContainer()` methods will
create an instance of a `Loader` class object, where the logger defined in the service client props is passed in

> as an optional parameter, `ILoaderProps.logger`, to the `Loader` constructor.

Reviewer 2: Are the details of what objects are created and how the logger is passed through relevant to consumers?

Author: from an implementation standpoint, no. This more or less just serves as providing slightly more context to the developers. But I agree this isn't diving deep enough to provide any meaningful background, and even more so now that logger is extracted away from `createContainer` and `getContainer`

Author: I will remove the details

Like demonstrated here, it is imperative to ensure `send()` is ultimately called at the end of custom properties for the
information to be piped to the container's telemetry system and sends the telemetry event.

> Another point worth noting is that with the customized logics, it is a great place where you can send these events to your external telemetry system to have them logged.

Author: This is where I'm pointing out where in the implementation the developer would send events to their telemetry endpoint. Not sure if this is enough or need more elaboration.

Reviewer: this callout should be below, where the doc covers a custom itelenetrybaselogger implementation.  when you put it here it's not clear what "it" in the callout refers to

Author: I was referring to having it in the `send()`, but I agree having it in the bottom section makes more sense.

});
```

> Another point worth noting is that with the customized logics, it is a great place where you can send these events to
> your external telemetry system to have them logged.

Reviewer 2: ```suggestion
Because you can customize the logic, you can send these events to your own external telemetry system to have them logged. This enables you to integrate Fluid Framework logging with your other telemetry.
```

}
```

> This custom logger should be provided in the service client constructor and will be created when creating or getting

Reviewer 2: ```suggestion
This custom logger should be provided in the service client constructor. Fluid will create an instance of the logger when creating or getting
```

> the container. Once the container is returned, the custom logger is now hooked up to the container's telemetry system.

Reviewer 2: ```suggestion
the container. The custom logger is hooked up to the container's telemetry system by the time the container is returned.
```

```ts
async function start(): Promise<void> {
  // Create a custom ITelemetryBaseLogger object to pass into the Tinylicious container
  // and hook to the Telemetry system
  const azureConfig = {
      connection: connectionProps,
      logger: new ConsoleLogger(),
  };
  const client = new AzureClient(azureConfig);
  let container: FluidContainer;
  let services: AzureContainerServices;

  // Get or create the document depending if we are running through the create new flow
  const createNew = !location.hash;
  if (createNew) {
      // The client will create a new detached container using the schema
      // A detached container will enable the app to modify the container before attaching it to the client
      {container, services} = await client.createContainer(containerSchema);
  }
```

Now, whenever a telemetry event is encountered, the custom `send()` method gets called and will print out the entire
event object.

<img src="https://fluidframework.blob.core.windows.net/static/images/consoleLogger_telemetry_in_action.png" alt="The
  ConsoleLogger sends telemetry events to the browser console for display.">

{{% callout warning %}}

> The purpose of `ConsoleLogger` is purely to demonstrate how the `ITelemetryBaseLogger` interface should be implemented. The
> Fluid Framework provides `DebugLogger` by default if a custom logger is not provided. For more information, please refer to
> [Using DebugLogger](#using-debuglogger) below instead of implementing something similar to `ConsoleLogger`.

Reviewer 2: ```suggestion
The purpose of `ConsoleLogger` is to demonstrate how the `ITelemetryBaseLogger` interface should be implemented. In typical usage, developers should instead use the `DebugLogger`, which is provided by default by the Fluid Framework. See [Using DebugLogger](#using-debuglogger) below instead of implementing something similar to `ConsoleLogger`.
```

{{% /callout %}}

### Using `DebugLogger`

> The `DebugLogger` offers a convenient way to output all telemetry events to the console. Like mentioned previously, creating
> or getting a container creates a `Loader` object, which then also provides a `DebugLogger` as a mixin in its constructor. Hence,
> `DebugLogger` is already present by default when creating/getting a container, and no extra steps are required to use it.

Reviewer 2: ```suggestion
The `DebugLogger` offers a convenient way to output all telemetry events to the console. `DebugLogger` is present by default when creating/getting a container, and no extra steps are required to use it.
```

I think the details of the code flow aren't relevant, so just focus on the usage.

Author: I agree, corresponding to the previous comment you mentioned above.

> Under the hood, `DebugLogger` uses the [debug](https://github.com/visionmedia/debug) library. `Debug` allows a library to log
> messages to a namespace. By default these messages aren't displayed but can be enabled by the app that is using the library. For
> our node apps enabling library logging can be done by setting the DEBUG environment variable - i.e.
> 
> In the browser you can enable them by setting `localStorage.debug` variable, after which you will need to reload the page.
> 
> `localStorage.debug = 'fluid:*'`
> 

Reviewer 2: ```suggestion
```suggestion
Under the hood, `DebugLogger` uses the [debug](https://github.com/visionmedia/debug) library. The `debug` library enables Fluid to send to a unique 'namespace,' `fluid`. By default these messages are hidden but they can be enabled
in both Node.js and a web browser.

**To enable Fluid Framework logging in the browser,** set the `localStorage.debug` variable in the JavaScript console, after which you will need to reload the page.

\`\`\`js
localStorage.debug = 'fluid:*'
\`\`\`

**To enable Fluid Framework logging in a Node.js application,** set the `DEBUG` environment variable when running the app.

```

I think you're missing the code sample for node.

}
```

> This custom logger is then created in the `azureConfig` before being passed into `AzureClient`

Reviewer: the config doesn't create the logger; it's created separately and provided as part of the config

Author: good point. I'll rephrase it 😊

constructor. The custom logger is now hooked up to the container's telemetry system.

```ts
async function start(): Promise<void> {
  // Create a custom ITelemetryBaseLogger object to pass into the Tinylicious container
  // and hook to the Telemetry system
  const azureConfig = {
      connection: connectionProps,
      logger: new ConsoleLogger(),
  };
  const client = new AzureClient(azureConfig);
  let container: FluidContainer;
  let services: AzureContainerServices;

>   let id: string;

Reviewer: nit: unused here

Author: Good point. I was thinking about copying the entire code snippet as is to maintain it's originality, but I realized the readers probably don't know where this code is originally from, so better to just focus on what's relevant 😅

  // Get or create the document depending if we are running through the create new flow
  const createNew = !location.hash;
  if (createNew) {
      // The client will create a new detached container using the schema
      // A detached container will enable the app to modify the container before attaching it to the client

>       ({container, services} = await client.createContainer(containerSchema));

Reviewer: extra parentheses are unnecessary here

Reviewer 3: I think these parentheses are required now that we don't send back a typed object, and declare `container` and `services` earlier
Agree though that the syntax is a bit weird

```

Now, whenever a telemetry event is encountered, the custom `send()` method gets called and will print out the entire
event object.

<img src="https://fluidframework.blob.core.windows.net/static/images/consoleLogger_telemetry_in_action.png" alt="The
  ConsoleLogger sends telemetry events to the browser console for display.">

{{% callout warning %}}

> The purpose of `ConsoleLogger` is purely to demonstrate how the `logger` interface should be implemented. The Fluid Framework provides `DebugLogger` by default if custom logger is not provided. For more information, please refer to [Using DebugLogger](#using-debuglogger) below instead of implementing something similar to `ConsoleLogger`.

Reviewer: ```suggestion
The purpose of `ConsoleLogger` is purely to demonstrate how the `ITelemetryBaseLogger` interface should be implemented. The Fluid Framework provides `DebugLogger` by default if a custom logger is not provided. For more information, please refer to [Using DebugLogger](#using-debuglogger) below instead of implementing something similar to `ConsoleLogger`.
```

{{% /callout %}}

### Using `DebugLogger`

> The `DebugLogger` offers a conveneient way to output all telemetry events to the console. Like mentioned previously, creating or getting a container creates a `Loader` object, which will then create a `DebugLogger` in its constructor. Hence, `DebugLogger` is already provided by default when creating/getting a container, and no extra steps is required to use the logger.

Reviewer: ```suggestion
The `DebugLogger` offers a convenient way to output all telemetry events to the console. Like mentioned previously, creating or getting a container creates a `Loader` object, which then also provides a `DebugLogger` as a mixin in its constructor. Hence, `DebugLogger` is already present by default when creating/getting a container, and no extra steps are required to use it.
```

> Under the hood, `DebugLogger` make use of the [debug](https://github.com/visionmedia/debug) library. `Debug` allows a library to log messages to a namespace. By default these messages aren't displayed but can be enabled by the app that is making use of the library. Debug is a popular package used by most major node modules (express, socket.io, etc...). For our node apps enabling library logging is as simple as setting the DEBUG environment variable - i.e.

Reviewer: ```suggestion
Under the hood, `DebugLogger` uses the [debug](https://github.com/visionmedia/debug) library. `Debug` allows a library to log messages to a namespace. By default these messages aren't displayed but can be enabled by the app that is using the library. For our node apps enabling library logging can be done by setting the DEBUG environment variable - i.e.
```

`DEBUG=fluid:*,connect:compress,connect:session`

> This is already done in our docker compose files for our own internal libraries which are all under the routerlicous namespace.

Reviewer 3: Not sure if this line is necessary and may just confuse people since we don't discuss routerlicious elsewhere here

Author: yup! I took it out from @heliocliu previous suggestion 😊

> In the browser you can enable them by setting localStorage.debug variable.
> 
> `localStorage.debug = 'fluid:*'`
> 
> After which you will need to reload the page.

Reviewer: ```suggestion
In the browser you can enable them by setting localStorage.debug variable, after which you will need to reload the page.

`localStorage.debug = 'fluid:*'`
```
